MoreLinq.Test/TransposeTest.cs

```diff
@@ -0,0 +1,232 @@
+        {
+            var matrix = new[]
+            {
+                new[] { 10, 11 },
+                new[] { 20 },
+                new int[0],
+                new[] { 30, 31, 32 }
+            };
+
+            var traspose = matrix.Transpose();
+
+            Assert.That(matrix.Last().Last(), Is.EqualTo(traspose.Last().Last()));
+            Assert.That(matrix.First().First(), Is.EqualTo(traspose.First().First()));
+        }
+
+        [Test]
+        public void TransposeWithAllInnerSequencesInfinite()
+        {
+            var matrix = MoreEnumerable.Generate(1, x => x + 1)
+                                       .Where(x => IsPrime(x))
+                                       .Take(3)
+                                       .Select(x => MoreEnumerable.Generate(x, n => n * x));
+
+            var result = matrix.Transpose().Take(5);
+
+            var expectations = new[]
+            {
+                new[] { 2,    3,    5 },
+                new[] { 4,    9,   25 },
+                new[] { 8,   27,  125 },
+                new[] { 16,  81,  625 },
+                new[] { 32, 243, 3125 }
+            };
+
+            AssertMatrix(result, expectations);
+        }
+
+        [Test]
+        public void TransposeWithSomeSequencesInfinite()
+        {
+            var matrix = MoreEnumerable.Generate(1, x => x + 1)
+                                       .Where(x => IsPrime(x))
+                                       .Take(3)
+                                       .Select((x, i) => i == 1 ? MoreEnumerable.Generate(x, n => n * x).Take(2)
+                                                                : MoreEnumerable.Generate(x, n => n * x));
+
+            var result = matrix.Transpose().Take(5);
+
+            var expectations = new[]
+            {
+                new[] { 2,    3,    5 },
+                new[] { 4,    9,   25 },
+                new[] { 8,        125 },
+                new[] { 16,       625 },
+                new[] { 32,      3125 }
+            };
+
+            AssertMatrix(result, expectations);
+        }
+
+        [Test]
+        public void TransposeOrderSequencesAreIteratedIsIrrelevant()
+        {
+            var matrix = new[]
+            {
+                new[] { 10, 11 },
+                new[] { 20 },
+                new int[0],
+                new[] { 30, 31, 32 }
+            };
+
+            var transpose = matrix.Transpose().ToList();
+
+            transpose[1].AssertSequenceEqual(11, 31);
+            transpose[0].AssertSequenceEqual(10, 20, 30);
+            transpose[2].AssertSequenceEqual(32);
+        }
+
+        [Test]
+        public void TransposeInnerSequencesAreConsumedLazies()
+        {
+            var matrix = new[]
+            {
+                MoreEnumerable.From(() => 10, () => 11),
+                MoreEnumerable.From(() => 20, () => 22),
+                MoreEnumerable.From(() => 30, () => throw new InvalidOperationException(), () => 31),
```

Comment:
Throw [`TestException`](https://github.com/morelinq/MoreLINQ/blob/6a170a021f996be9908f0cd8417cad273d8765da/MoreLinq.Test/TestException.cs) (added in 6a170a021f996be9908f0cd8417cad273d8765da) instead and catch the same later during assertion.